toc/rfc/rfc-draft-ssh-over-ws.md

```diff
@@ -0,0 +1,270 @@
+		User: fmt.Sprintf("cf:%s/%d", c.app.GUID, opts.Index),
+		Auth: []ssh.AuthMethod{
+			ssh.Password(c.token),
+		},
+		HostKeyCallback: fingerprintCallback(opts, c.sshEndpointFingerprint),
+	}
+
+	// Wrap SSH in WebSocket if possible
+	var secureClient SecureClient
+	if c.sshWsEndpoint != "" {
+		secureClient, err = tunnelSSHThruWebSocket(c.sshWsEndpoint, clientConfig)
+	}
+
+	if err != nil {
+		// Fall back to SSH over TCP
+		secureClient, err = c.secureDialer.Dial("tcp", c.sshEndpoint, clientConfig)
+		if err != nil {
+			return err
+		}
+	}
+
+	c.secureClient = secureClient
+	c.opts = opts
+	return nil
+}
+```
+Code adapted from [cloudfoundry/cli ssh.go](https://github.com/cloudfoundry/cli/blob/main/cf/ssh/ssh.go#L119)
+
+The `sshWsEndpoint` would have to be fetched from the `app_ssh_ws_endpoint` field of the "info" endpoint of CAPI.
+
+
+A `tunnelSSHThruWebSocket` function could look something like this:
+```
+func tunnelSSHThruWebSocket(url string, config *ssh.ClientConfig) (SecureClient, error) {
+	wsConn, err := websocket.Dial(url, "", url)
+	if err != nil {
+		return nil, err
+	}
+
+	c, chans, reqs, err := ssh.NewClientConn(wsConn, url, config)
+	if err != nil {
+		return nil, err
+	}
+	sshClient := ssh.NewClient(c, chans, reqs)
+	return &secureClient{client: sshClient}, nil
+}
+```
+
+- The rest of CF CLI SHOULD be unchanged and work as before.
+
+#### CAPI
+
+- CAPI MAY provide an additional `app_ssh_ws_endpoint` field in the "v2/info" endpoint:
+```
+  "app_ssh_ws_endpoint": "wss://ssh.cf-app.com"
+```
+
+- The URL presented in the field MUST be `wss://` + the route as announced by route-registrar for SSH proxy (see [CF Deployment](#cf-deployment))
+
+#### SSH Proxy
+
+- Diego SSH Proxy SHOULD provide the option to [launch a separate WebSocket server](https://github.com/cloudfoundry/diego-ssh/blob/main/cmd/ssh-proxy/main.go#L65) besides the regular SSH server
+- SSH Proxy SHOULD provide a separate handler for unwrapping WebSocket connections:
+```
+	sshWsProxy := wsProxy.New(logger, proxySSHServerConfig, metronClient, tlsConfig)
+	wsServer := server.NewServer(logger, sshProxyConfig.Address, websocket.Handler(sshWsProxy), time.Duration(sshProxyConfig.IdleConnectionTimeout))
+```
+Changes in [ssh-proxy/main.go](https://github.com/cloudfoundry/diego-ssh/blob/main/cmd/ssh-proxy/main.go#L65C1-L67C1)
+
+```
+func (p *WsProxy) HandleConnection(wsConn *websocket.Conn) {
+	logger := p.logger.Session("handle-ws-connection")
+	defer wsConn.Close()
+
+    // Here we start a SSH handshake inside the WebSocket tunnel
+	serverConn, serverChannels, serverRequests, err := ssh.NewServerConn(wsConn, p.serverConfig)
+
+(...)
+```
+Changes in [proxy.go](https://github.com/cloudfoundry/diego-ssh/blob/main/proxy/proxy.go#L67)
+
+**Remark:**
+(these changes are not meant as verbatim code; they only illustrate areas that would need to change)
+
+
+- SSH Proxy MUST retain the legacy SSH server on port 2222 for backwards compatibiliy
```

Comment:
As noted, I would see an opt-in to disable this port.

Reply:
Agreed as above changing to SHOULD. I will also add a statement that the code for the legacy server shall not be removed.